### PR TITLE
feat(modal): add parentSelector prop, use portals

### PR DESCRIPTION
**Pull request by Mjloturco**
Modal currently does not render on bootstrap pages, because paragon uses class names different from the default bootstrap class names. This PR brings the modal more in line with bootstrap's default modal and renders the modal via [```createPortal()```](https://reactjs.org/docs/portals.html)

### CODE REVIEW

**Comment by Mjloturco** [comment]
@reviewers, just noticed that it looks like a bunch of build changes were pulled into the diff. I'll look to remove those before merging anything.

File: src/Modal/index.jsx [hunk 1]
```diff
@@ -43,6 +58,10 @@ class Modal extends React.Component {
     }
   }
 
+  componentWillUnmount() {
+    modalRoot.removeChild(this.el);
```
Comment (jaebradley) [inline src/Modal/index.jsx]: Just out of curiosity are we missing anything that we'd get from [`ReactDOM.unmountComponentAtNode`](https://reactjs.org/docs/react-dom.html#unmountcomponentatnode)?

File: src/Modal/index.jsx [hunk 2]
```diff
@@ -20,12 +26,21 @@ class Modal extends React.Component {
 
     this.headerId = newId();
 
+    this.el = document.createElement('div');
+
+    if (!modalRoot) {
+      modalRoot = document.createElement('div');
+      modalRoot.id = modalRootId;
+      document.body.appendChild(modalRoot);
```
Comment (jaebradley) [inline src/Modal/index.jsx]: What happens if I want to append this modal to a specific parent in the DOM? Should we add a `parentSelector` prop that defaults to the document body?

File: src/Modal/index.jsx [hunk 3]
```diff
@@ -20,12 +26,21 @@ class Modal extends React.Component {
...
+    }
+
     this.state = {
       open: props.open,
     };
   }
 
   componentDidMount() {
+    modalRoot.appendChild(this.el);
```
Comment (jaebradley) [inline src/Modal/index.jsx]: So I'm probably totally missing something here...

It looks like we're creating a `modalRoot` `div` that we're appending to document body. Then for this `modalRoot` `div` we're appending another `div` (`this.el`), which is where we're creating the portal.

Is there any reason we're not appending `this.el` to document body? In other words, what is `modalRoot` doing that `this.el` couldn't do?

File: src/Modal/index.jsx [hunk 4]
```diff
@@ -9,6 +10,11 @@ import Icon from '../Icon';
 import newId from '../utils/newId';
 import Variant from '../utils/constants';
 
+
+const modalRootId = 'modal-root';
+let modalRoot = document.getElementById(modalRootId);
```
Comment (jaebradley) [inline src/Modal/index.jsx]: Having this outside the scope of Modal makes me nervous - what happens if somebody deletes this element in the DOM? `modalRoot` will still probably be defined right?

File: src/Modal/index.jsx (hunk 2 above)
Comment (Mjloturco) [reply src/Modal/index.jsx]: We could add a parentSelector prop, though I believe we do expect to want to always render the modal at the top level component, that was at a suggestion from @arizzitano who may have more thoughts on that.

File: src/Modal/index.jsx (hunk 1 above)
Comment (Mjloturco) [reply src/Modal/index.jsx]: I'm not sure, the docs for unmountComponentAtNode specifically mentions cleaning up eventListeners and such, which removeChild does not. I'll check to see if unmountComponentAtNode accomplishes what we want.

**Comment by Mjloturco** [comment]
@fysheets hi

**Comment by jaebradley** [comment]
Just because I have a terrible memory and have already forgotten the reason for these changes here's a relevant conversation from the FedX channel

image

Just want to add this for historical purposes.

File: src/Modal/Modal.stories.jsx [hunk 5]
```diff
@@ -50,6 +51,11 @@ class ModalWrapper extends React.Component {
 ModalWrapper.propTypes = {
   title: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
   body: PropTypes.oneOfType([PropTypes.string, PropTypes.element]).isRequired,
+  parentSelector: PropTypes.string,
+};
+
+ModalWrapper.defaultProps = {
+  parentSelector: '',
```
Comment (jaebradley) [inline src/Modal/Modal.stories.jsx]: I'm being uber-paranoid, but can we make the default value 'body' or something similar?

Comment (fysheets) [reply src/Modal/Modal.stories.jsx]: screen shot 2018-03-07 at 4 27 39 pm
Seconded :)

**Comment by fysheets** [comment]
Unblocking from my end - as we've discussed and I've tested to get a plan of attack from our team's end. Will let others continue with additional technical reviews etc. Also, obviously pending tests :)

File: src/Modal/index.jsx [hunk 6]
```diff
@@ -43,6 +63,10 @@ class Modal extends React.Component {
     }
   }
 
+  componentWillUnmount() {
+    ReactDOM.unmountComponentAtNode(this.el);
```
Comment (jaebradley) [inline src/Modal/index.jsx]: Does `modal-root` need to get unmounted as well?

File: src/Modal/index.jsx [hunk 7]
```diff
@@ -20,12 +22,30 @@ class Modal extends React.Component {
 
     this.headerId = newId();
 
+    const modalRootId = 'modal-root';
+    this.modalRoot = document.getElementById(modalRootId);
```
Comment (jaebradley) [inline src/Modal/index.jsx]: What happens if there are two modals on the same page if `id`s are supposed to be unique across the DOM?

**Pull request merged**